docs/STRUCTURE-GUIDELINES.md

```diff
@@ -0,0 +1,233 @@
+
+Before getting into the specific directory/folde structure, let's review a few premises driving that structure.
+
+### A sample can be composed by multiple applications
+
+An important characteristic of the ML.NET samples is that some/many of them might have multiple related apps such as a console app for model creation/training and a different app for deploying the trained model, such as a Web app or desktop app. Each of those apps might be very different in purpose (ML model workflow vs. production app predicting with a trained model), therefore, the folder structure can and must be different depending on the type of app.
+
+### Application projects should be autonomous in structure
+
+The main goal here is that you should be able to copy the app's folder (such a web app folder) to any server or dev machine and still have everything it needs to run. That includes:
+
+ - For an "end-user" app (app predicting with a trained model), its folder must include the ML.NET model file it needs for predicting.
+ - For a "model builder/training" app, its folder must include the training and test datasets or if those files are too big for GitHub, then it should implement an automated way for download those files. 
+
+Essentially, the rule is that everything you need to run each app must be available within the app's project folder, without any absolute filepaths or references to other folders in the filesystem out of the scope of the app. 
+
+This rule can be arguable for your own ML workflow if you want to be more agile in your own apps by accessing common folders for generated models. However, for samples that must be able to run out-of-the-box when you just copy a single app (i.e. only the "production" app) to a different machine/server, this rule should be followed when implementing the ML.NET samples.
+
+### Directory structure for a sample with multiple apps
+
+Note that "Model-Builder-App" or "Production-App" would have your own naming. It is the concept the important point.
+
+```
+ML.NET Sample
+├── Solution.sln               <- Top-level VS solution targeting all apps in the sample
+├── README.md                  <- Top-level README for developers
+├── docs                       <- Global docs folder documenting the whole sample
+├── images                     <- Images used by the .md documents
+│
+├── Model-Builder-App          <- Usually a console app for building/training/evaluating/testing a model
+│    ├── Data
+│    │   ├── Raw               <- The original, immutable data dump.
+│    │   ├── Working           <- Intermediate data that has been transformed.
+│    │   └── Processed         <- The final, canonical datasets for training/evaluating/testing.
+│    ├── Models                <- The trained, generated saved/serialized ML.NET model files
+│    ├── MyConsoleApp          <- Source code for this particular app
```

Comment:
>MyConsoleApp [](start = 9, length = 12)

Is this extra level needed? It's not clear which of Model-Builder-App or MyConsoleApp is the project folder. (Project folders need to be called out)

Reply:
I think we do. There are three concepts here. 
- 1. The logical end to end sample/application (e.g. CustomersSegmentation). I our case, this is a folder for the sample. If it were a real application, this would be a whole GitHub repo.
- 2. The specific logical App, either MLModelBuilder or the End-User-App. This includes assets like datasets, generated MLModels, or even databases, etc. out of the source-code folder.
- 3. The VS-code project , for instance:
  - CustomersSegmentation-console-app
  - Customers-Dasboard-web-app